### PR TITLE
Add tournament streaming and management toolchain

### PR DESCRIPTION
**Pull request by peppy**
PRing to potentially get some feedback on this. Is a big change but only going to grow over the week as I am for functional completion to replace current tournament streaming UI in preparation for OWC.

Remaining tasks:

- [x] Add date entry
- Fix zoom logic
- Move as much event bind/reactive logic to non-drawable classes as possible
- [x] Add proper non-visualtests project and remove menu shortcut for drawings
- [x] Depends on ppy/osu-framework#2547.
- [x] Depends on #5052.
- [x] Depends on #5054.

~Sample bracket: https://gist.github.com/peppy/0523577997de3c698388c11152e62f60~ (outdated format)

### CODE REVIEW

File: osu.Game/Overlays/Settings/SettingsItem.cs [hunk 1]
```diff
@@ -63,6 +63,7 @@ public virtual Bindable<T> Bindable
 
             set
             {
+                controlWithCurrent?.Current.UnbindBindings();
```
Comment (smoogipoo) [inline osu.Game/Overlays/Settings/SettingsItem.cs]:
```suggestion
                if (bindable != null)
                    controlWithCurrent?.Current.UnbindFrom(bindable);
```

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game.Tournament/Components/DateTextBox.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        private Bindable<DateTimeOffset> bindable;
+
+        public DateTextBox()
+        {
+            base.Bindable = new Bindable<string>();
+            ((OsuTextBox)Control).OnCommit = (sender, newText) =>
+            {
+                try
+                {
+                    bindable.Value = DateTimeOffset.Parse(sender.Text);
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tournament/Components/DateTextBox.cs]: newText is a bool here

File: osu.Game.Tournament/Components/DateTextBox.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                }
+                catch
+                {
+                    bindable.TriggerChange();
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tournament/Components/DateTextBox.cs]: This resets the text to its previous value on a fail to parse. Will add comment.

File: osu.Game.Tournament/Components/TournamentBeatmapPanel.cs [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                    Colour = Color4.Gray,
+                    Blending = BlendingMode.Additive,
+                    Alpha = 0,
+                },
+            });
+
+            if (!string.IsNullOrEmpty(mods))
+                AddInternal(new Sprite
+                {
+                    Texture = new LargeTextureStore(new TextureLoaderStore(new StorageBackedResourceStore(storage))).Get($"mods/{mods}"),
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tournament/Components/TournamentBeatmapPanel.cs]: no idea!

File: osu.Game.Tournament/Screens/Ladder/Components/DrawableMatchTeam.cs [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+        private readonly Func<bool> isWinner;
+        private LadderEditorScreen ladderEditor;
+
+        [Resolved]
+        private LadderInfo ladderInfo { get; set; }
+
+        private void setCurrent()
+        {
+            //todo: tournamentgamebase?
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tournament/Screens/Ladder/Components/DrawableMatchTeam.cs]: no, still valid. i wanted to remove a lot of the action flow from the Drawable classes. still haven't been done yet.

File: osu.Game.Tournament/Screens/Ladder/Components/MatchPairing.cs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        public readonly Bindable<MatchPairing> LosersProgression = new Bindable<MatchPairing>();
+
+        /// <summary>
+        /// Should not be set directly. Use LadderInfo.CurrentMatch.Value = this instead.
+        /// </summary>
+        public readonly Bindable<bool> Current = new Bindable<bool>();
+
+        public readonly Bindable<DateTimeOffset> Date = new Bindable<DateTimeOffset>();
+
+        [JsonProperty]
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tournament/Screens/Ladder/Components/MatchPairing.cs]: was to get around this:
image

open to suggestion (it's only read from serialisation currently; no in-game interface

File: osu.Game.Tournament/Screens/Ladder/Components/TournamentGrouping.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+{
+    [Serializable]
+    public class TournamentGrouping
+    {
+        public readonly Bindable<string> Name = new Bindable<string>();
+        public readonly Bindable<string> Description = new Bindable<string>();
+
+        public readonly BindableInt BestOf = new BindableInt(9) { Default = 9, MinValue = 3, MaxValue = 23 };
+
+        [JsonProperty]
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tournament/Screens/Ladder/Components/TournamentGrouping.cs]: Same reasoning as above. Shall I add it to all the other properties to match?

File: osu.Game.Tournament/Screens/Ladder/LadderEditorScreen.cs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    [Cached]
+    public class LadderEditorScreen : LadderScreen, IHasContextMenu
+    {
+        [Cached]
+        private LadderEditorInfo editorInfo = new LadderEditorInfo();
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            ((Container)InternalChild).Add(new LadderEditorSettings
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tournament/Screens/Ladder/LadderEditorScreen.cs]: This is specifically adding to the container created by `LadderScreen`. ie. a Composite's `Content`. I'll name and expose as Content.

File: osu.Game.Tournament/Screens/Ladder/LadderEditorScreen.cs [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+                };
+            }
+        }
+
+        public void Remove(MatchPairing pairing)
+        {
+            PairingsContainer.FirstOrDefault(p => p.Pairing == pairing)?.Remove();
+        }
+
+        private class JoinRequestHandler : CompositeDrawable
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tournament/Screens/Ladder/LadderEditorScreen.cs]: it visualises and creates a result (the click logic results in a join being fulfilled). i'm fine with renaming though.

**Comment by peppy** [comment]
This is mostly ready to go now.

Changes made:
- Cleaned up bindable flow in many areas
- Made sure the tools can be used from scratch
- Added UI editing of all elements (except Team's players, coming soon)
- Improved render and invalidation logic of main bracket display

These are ready to be used be non-world-cup tournament for an enthusiast user. There's plenty of UX improvements to be applied, but I'm quite happy for an initial release.

**Comment by peppy** [comment]
Added player editor and fixed startup issues when videos are not present.

**Comment by peppy** [comment]
- Added round-beatmap editor.
- Made a very easily extensible base class for editor screens.
- Added ability to pre-populate teams with countries.
- Added ability to clear out all teams or rounds quickly.
- Extracted all `OsuGame` changes to dependent PRs (there should now be zero line added, and only `Drawings` screen extraction resulting in removals)
- Renamed classes to be more logical (`MatchPairing` -> `TournamentMatch`; `Grouping` -> `TournamentRound`).
- Added folder grouping to rider run configurations (they were getting cluttered).
- Added folders to tournament visual tests and added any missing tests.

@smoogipoo i'm seriously done on this now. Any further changes will be in follow-up PRs.

**Comment by smoogipoo** [comment]
@peppy I've pushed a few fixed, there's nothing super bad here and can be cleaned up later. See if you agree with my changes.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**